src/components/Rating/Rating.tsx

```diff
@@ -0,0 +1,110 @@
+import { useState } from "react";
+import "./rating.scss"; // Assuming you have a Sass file for styling
+
+
+export interface RatingProps {
+    totalStars: number;
+    shape?: string;
+    disabled?: boolean
+    readOnly?: boolean;
+    readOnlyValue?: number
+}
+
+const Rating: React.FC<RatingProps> = ({
+    totalStars,
+    shape = '\u2605',
+    disabled = false,
+    readOnly = false,
+    readOnlyValue = 5
+}) => {
+    const defaultRating = readOnly ? readOnlyValue : 0;
+
+    const numberOfStars = totalStars
+    const [rating, setRating] = useState(defaultRating);
+    const [hoveredRating, setHoveredRating] = useState(0);
+
+    const handleStarClick = () => {
+        if (!disabled && !readOnly) {
+            setRating(hoveredRating);
+
+        }
+    };
+
+    const handleStarHover = (e: React.MouseEvent<HTMLSpanElement>, index: number) => {
+        if (!disabled && !readOnly) {
+            const starWidth = (e.target as HTMLSpanElement).offsetWidth;
```

Comment:
Strange that you have to typecast here despite typing the event... I wonder how you could avoid this.